src/pil/backend/powdr_pil.rs

```diff
@@ -0,0 +1,488 @@
+            // need to convert lookups.
+            generate_pil_lookups(pil, pil_ir, step_type, step_type_name, annotations_map);
+        }
+    }
+}
+
+fn generate_pil_constraints<F: Debug + Clone>(
+    pil: &mut String,
+    pil_ir: &PilCircuit<F>,
+    step_type: &UUID,
+    step_type_name: String,
+    annotations_map: &HashMap<UUID, String>,
+) {
+    if let Some(constraints) = pil_ir.constraints.get(step_type) {
+        for expr in constraints {
+            // `convert_to_pil_expr_string` recursively converts expressions to PIL
+            // strings, using standardized variable names from
+            // `annotations_map`.
+            let expr_string = convert_to_pil_expr_string(expr.clone(), annotations_map);
+            // Each constraint is in the format of `step_type * constraint = 0`, where
+            // `step_type` is a fixed step selector column and
+            // `constraint` the actual constraint expression.
+            writeln!(pil, "{} * {} = 0;", step_type_name, expr_string).unwrap();
+        }
+    }
+}
+
+fn generate_pil_transitions<F: Debug + Clone>(
+    pil: &mut String,
+    pil_ir: &PilCircuit<F>,
+    step_type: &UUID,
+    step_type_name: String,
+    annotations_map: &HashMap<UUID, String>,
+) {
+    let is_last_step_instance = 1
+        == *pil_ir
+            .step_types_instantiations
+            .get(step_type)
+            .unwrap()
+            .last()
+            .unwrap();
+
+    if let Some(transitions) = pil_ir.transitions.get(step_type) {
+        for expr in transitions {
+            let expr_string = convert_to_pil_expr_string(expr.clone(), annotations_map);
+            // Disable transition constraint in the last step instance.
+            if is_last_step_instance {
+                writeln!(
+                    pil,
+                    "(1 - ISLAST) * {} * {} = 0;",
+                    step_type_name, expr_string
+                )
+                .unwrap();
+            } else {
+                writeln!(pil, "{} * {} = 0;", step_type_name, expr_string).unwrap();
+            }
+        }
+    }
+}
+
+fn generate_pil_lookups<F: Debug + Clone>(
+    pil: &mut String,
+    pil_ir: &PilCircuit<F>,
+    step_type: &UUID,
+    step_type_name: String,
+    annotations_map: &HashMap<UUID, String>,
+) {
+    if let Some(lookups) = pil_ir.lookups.get(step_type) {
+        for lookup in lookups {
+            let mut lookup_source: Vec<String> = Vec::new();
+            let mut lookup_destination: Vec<String> = Vec::new();
+            for (src, dest) in lookup {
+                lookup_source.push(convert_to_pil_expr_string(src.clone(), annotations_map));
+                lookup_destination.push(convert_to_pil_expr_string(dest.clone(), annotations_map));
+            }
+            // PIL lookups have the format of `step_type { lookup_sources } in {
+            // lookup_destinations }`.
+            writeln!(
```

Comment:
@qwang98 Do we have or we don't have lookups in PIL?

Reply:
We do have lookups (source to destination columns) in PIL but not lookup tables (destination columns only).